Create documentation that provides resource sizing for clusters of incremental scale (#1510)
spiffe/spire: doc/scaling_spire.md

 
Complimentary to scaling SPIRE servers horizontally to deal with load, a nested topology helps to segment failure domains.

## Federated SPIRE

A requirement or objective may exist to make use of multiple roots of trust: perhaps because an organization has different organizational divisions with different administrators, or because they have separate staging and production environments that occasionally need to communicate. A similar use-case is SPIFFE interoperability between organizations, such as between a cloud provider and its customers. Both use-cases require a well-defined, interoperable method for a workload in one trust domain to authenticate a workload in a different trust domain.
Trust between the different trust domains is established by authenticating the respective Bundle Endpoint. For additional detail on how this is achieved, refer to the following SPIFFE spec that describe the mechanism: 
https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#5-spiffe-bundle-endpoint

# Sizing Clusters

> This table provides a sample reference for calculating resource allocation for SPIRE servers SPIRE deployments. It displays for different combinations of "Number of Workloads" and "Number of Agents", a corresponding baseline SPIRE Server CPU/Memory configuration. These numbers are intended to serve as sample guidance, and are not representative of upper bound scale limits nor configuration maximums. The number of workloads and agents shown do not represent the theoretically possible SPIRE deployment scale.

Reviewer: Suggest replacing:
"These numbers are intended to serve as sample guidance, and are not representative of upper bound scale limits nor configuration maximums. The number of workloads and agents shown do not represent the theoretically possible SPIRE deployment scale."

With something stronger, like:

These reference numbers are based on experiments conducted in a testing environment.  They are meant as order-of-magnitude guidelines only, and do not represent a performance guarantee for any particular user environment.   Note that datastore performance is not addressed here, and can potentially limit SPIRE performance.  There are many factors that will influence SPIRE performance including, but not limited to:

- SVID and root certificate TTLs
- Number and distribution of workloads per node
- Heavy JWT-SVID use (because JWTs must be signed as needed, rather than pre-stashed like x509s)
- Frequency of registration changes
- Other processes running on a SPIRE Server node

Author: Replaced with the suggestion.

It is important to note when scaling the number servers horizontally, be it for high availability or load distribution purposes, that in any configuration of two or more SPIRE servers, all servers share the same datastore, while each individual server maintains an independent root certificate/key. 

# Topologies

There are three main SPIRE deployment topologies; single trust domain, nested SPIRE, and federated SPIRE. Multiple considerations should are to be factored in when choosing one topology design over another.

A common motivation for a single overarching trust domain is to issue identifies from a single Certificate Authority, as it reduces the number of components to manage. However, when deploying a single SPIRE trust domain to span regions and clouds. 

> Managing a shared datastore across geographically dispersed locations or across cloud providers may represent a challenge. In such circumstances, a solution is to configure SPIRE servers can be configured in Nested topology.

Author: I rewrote the two sentences for clarity.

## Nested SPIRE

> Nested SPIRE allows SPIRE servers to be chained together, and for all servers to still issue identities in the same trust domain. It works by co-locating an agent with downstream servers. Each downstream server loads the CA credentials from the Workload API of the top or global SPIFFE implementation. The downstream node performs both node and workload attestation, which in turn allows automating the bring-up of new downstream servers. 

Reviewer 2: This description isn't quite accurate. The downstream SPIRE server obtains credentials over the Workload API that it uses to _directly_ authenticate with the upstream SPIRE server to obtain an intermediate CA.

Also, I couldn't quite grok the last sentence. What is the purpose of talking about node attestation here? I'm not following something :grimacing:

Author: Thanks. I added the order of operations here.

 
In this configuration, the top tier SPIRE servers hold the root certificate/key, and the downstream servers request an intermediate signing certificate to use as the server's X.509 signing authority. It provides for resilience as the top tier can go down, and intermediate certificates will continue to operate. 
 
Complimentary to scaling SPIRE servers horizontally to deal with load, a nested topology helps to segment failure domains.

## Federated SPIRE

A requirement or objective may exist to make use of multiple roots of trust: perhaps because an organization has different organizational divisions with different administrators, or because they have separate staging and production environments that occasionally need to communicate. A similar use-case is SPIFFE interoperability between organizations, such as between a cloud provider and its customers. Both use-cases require a well-defined, interoperable method for a workload in one trust domain to authenticate a workload in a different trust domain.

> Trust between the different trust domains is established by authenticating the respective Bundle Endpoint. For additional detail on how this is achieved, refer to the following SPIFFE spec that describe the mechanism: 

Reviewer 2: Authenticating the endpoint is the first step. Obtaining the foreign trust domain bundle via the authenticated endpoint is the next.

Author: Updated to outline the order of operations.

# Scalability

> A SPIRE deployment is composed of a number one or more SPIRE Servers that share a replicated datastore, and one or more SPIRE Agents. With only one SPIRE server per SPIRE deployment, the memory and CPU usage tends to grow proportionally with the number of workloads in the deployment. To support very large numbers of workloads and agents within a given trust domain (tens of thousands or hundreds of thousands of nodes), the number of SPIRE Servers can be scaled horizontally. With multiple servers, the load is distributed. In addition to capacity, multiple SPIRE Servers can be configured for high availability.

Reviewer 2: There is one spot where we say "SPIRE server" instead of "SPIRE Server". Suggest being consistent one way or the other.

> It is important to note when scaling the number servers horizontally, be it for high availability or load distribution purposes, that in any configuration of two or more SPIRE servers, all servers share the same datastore, while the root certificate/key might be shared (i.e. UpstreamAuthority), each server CA is is unique per server (either self-signed, or an intermediate off of the root).

Reviewer 2: ```suggestion
It is important to note when scaling the number servers horizontally, be it for high availability or load distribution purposes, that in any configuration of two or more SPIRE servers, all servers share the same datastore. Each server maintains its own CA however, which may be self-signed or an intermediate off of a shared root authority (i.e. when configured with an UpstreamAuthority).
```

# Topologies

> There are three main SPIRE deployment topologies; single trust domain, nested SPIRE, and federated SPIRE. Multiple considerations should are to be factored in when choosing one topology design over another.

Reviewer 2: ```suggestion
There are three main SPIRE deployment topologies; single trust domain, nested SPIRE, and federated SPIRE. Multiple factors should be considered when choosing one topology design over another.
```

> A common motivation for a single overarching trust domain is to issue identifies from a single Certificate Authority, as it reduces the number of components to manage. However, when deploying a single SPIRE trust domain to span regions, platforms, and cloud provider environments, there is a level of complexity associated with managing a shared datastore across geographically dispersed locations or across cloud provider boundaries. In such circumstances where a trust domain may stretch across technology boundaries, a solution to address the use of a shared datastore is to configure SPIRE servers in a nested topology.

Reviewer 2: ```suggestion
A common motivation for a single overarching trust domain is to issue identities from a single Certificate Authority, as it reduces the number of components to manage. However, when deploying a single SPIRE trust domain to span regions, platforms, and cloud provider environments, there is a level of complexity associated with managing a shared datastore across geographically dispersed locations or across cloud provider boundaries. Under these circumstances a solution to address the use of a shared datastore is to configure SPIRE servers in a nested topology.
```

## Nested SPIRE

Nested SPIRE allows SPIRE servers to be chained together, and for all servers to still issue identities in the same trust domain. It works by co-locating an agent with downstream servers. The downstream SPIRE server obtains credentials over the Workload API that it uses to directly authenticate with the upstream SPIRE server to obtain an intermediate CA. 
 
This topology is well suited for multi-cloud deployments. It can be reasoned as the top server being a global server (or set of servers for high availability). The downstream servers may be deployed at either regional or cluster level. Due to the ability to mix and match node attestors, the downstream servers can reside in different clouds. 
 

> In this configuration, the top tier SPIRE servers hold the root certificate/key, and the downstream servers request an intermediate signing certificate to use as the server's X.509 signing authority. It provides for resilience as the top tier can go down, and intermediate certificates will continue to operate. 

Reviewer 2: ```suggestion
In this configuration, the top tier SPIRE servers hold the root certificate/key, and the downstream servers request an intermediate signing certificate to use as the server's X.509 signing authority. It provides for resilience as the top tier can go down, and intermediate servers will continue to operate. 
```
